python/ray/util/ray_executor.py

```diff
@@ -0,0 +1,317 @@
+import itertools
+import time
+from concurrent.futures import Executor, Future
+from functools import partial
+from typing import (
+    Callable,
+    Optional,
+    List,
+    Any,
+    TypeVar,
+    ParamSpec,
+    Iterable,
+    Iterator,
+    TYPE_CHECKING,
+)
+
+import ray
+from ray.util.actor_pool import ActorPool
+from ray.util.annotations import PublicAPI
+
+
+# Typing -----------------------------------------------
+
+if TYPE_CHECKING:
+    from ray._private.worker import BaseContext, RemoteFunction0
+
+T = TypeVar("T")
+P = ParamSpec("P")
+
+# ------------------------------------------------------
+
+
+@PublicAPI(stability="alpha")
+class RayExecutor(Executor):
+    """`RayExecutor` is a drop-in replacement for `ProcessPoolExecutor` and
+    `ThreadPoolExecutor` from `concurrent.futures` but distributes and executes
+    the specified tasks over a Ray cluster instead of multiple processes or
+    threads.
+
+    Args:
+        max_workers:
+            If max_workers=None, the work is distributed over the number of
+            CPUs available in the cluster (num_cpus) , otherwise the number of
+            CPUs is limited to the value of max_workers.
+
+        All additional keyword arguments are passed to ray.init()
+        (see https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+
+        For example, this will connect to a cluster at the specified address:
+        .. code-block:: python
+
+            RayExecutor(address='192.168.0.123:25001')
+
+        Note: excluding an address will initialise a local Ray cluster.
+    """
+
+    def __init__(
+        self, max_workers: Optional[int] = None, shutdown_ray: bool = True, **kwargs
```

Comment:
To make the API simpler: we can remove the `shutdown_ray` option, users can always control whether they want to  call `ray.shutdown()` (in most cases, this doesn't need to be called since Ray will auto shutdown when process exits)